File: app/src/main/java/wangdaye/com/geometricweather/theme/weatherView/materialWeatherView/implementor/WindImplementor.java
package wangdaye.com.geometricweather.theme.weatherView.materialWeatherView.implementor;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;

import androidx.annotation.ColorInt;
import androidx.annotation.Size;
import androidx.core.graphics.ColorUtils;

import java.util.Random;

import wangdaye.com.geometricweather.theme.weatherView.materialWeatherView.MaterialWeatherView;

public class WindImplementor extends MaterialWeatherView.WeatherAnimationImplementor {

    private final Paint mPaint;
    private final Wind[] mWinds;

    private float mLastDisplayRate;

    private float mLastRotation3D;
    private static final float INITIAL_ROTATION_3D = 1000;

    private static final int WIND_COUNT = 240;

    @ColorInt
    private final int mBackgroundColor;

    private static class Wind {

        float x;
        float y;
        float width;
        float height;

        RectF rectF;
        float speed;

        @ColorInt
        int color;
        float scale;

        private final int mViewWidth;
        private final int mViewHeight;

        private final int mCanvasSize;

        private final float MAX_WIDTH;
        private final float MIN_WIDTH;
        private final float MAX_HEIGHT;
        private final float MIN_HEIGHT;

        private Wind(int viewWidth, int viewHeight, @ColorInt int color, float scale) {
            mViewWidth = viewWidth;
            mViewHeight = viewHeight;

            mCanvasSize = (int) Math.pow(
                    viewWidth * viewWidth + viewHeight * viewHeight,
                    0.5
            );

            this.rectF = new RectF();
            this.speed = (float) (
                    mCanvasSize / (
                            1000.0 * (0.5 + new Random().nextDouble())
                    ) * 4.0
            );
            this.color = color;
            this.scale = scale;

            this.MAX_HEIGHT = 0.006f * mCanvasSize;
            this.MIN_HEIGHT = 0.003f * mCanvasSize;
            this.MAX_WIDTH = this.MAX_HEIGHT * 10;
            this.MIN_WIDTH = this.MIN_HEIGHT * 6;

            init(true);
        }

        private void init(boolean firstTime) {
            Random r = new Random();
            y = r.nextInt(mCanvasSize);
            if (firstTime) {
                x = r.nextInt((int) (mCanvasSize - MAX_HEIGHT)) - mCanvasSize;
            } else {
                x = -MAX_HEIGHT;
            }
            width = MIN_WIDTH + r.nextFloat() * (MAX_WIDTH - MIN_WIDTH);
            height = MIN_HEIGHT + r.nextFloat() * (MAX_HEIGHT - MIN_HEIGHT);

            buildRectF();
        }

        private void buildRectF() {
            float x = (float) (this.x - (mCanvasSize - mViewWidth) * 0.5);
            float y = (float) (this.y - (mCanvasSize - mViewHeight) * 0.5);
            rectF.set(x, y, x + width * scale, y + height * scale);
        }

        void move(long interval, float deltaRotation3D) {
            x += speed * interval
                    * (Math.pow(scale, 1.5)
                    + 5 * Math.sin(deltaRotation3D * Math.PI / 180.0) * Math.cos(16 * Math.PI / 180.0));
            y -= speed * interval
                    * 5 * Math.sin(deltaRotation3D * Math.PI / 180.0) * Math.sin(16 * Math.PI / 180.0);

            if (x >= mCanvasSize) {
                init(false);
            } else {
                buildRectF();
            }
        }
    }

    public WindImplementor(@Size(2) int[] canvasSizes) {
        mPaint = new Paint();
        mPaint.setStyle(Paint.Style.FILL);
        mPaint.setAntiAlias(true);

        mBackgroundColor = Color.rgb(233, 158, 60);
        int[] colors = new int[] {
                Color.rgb(240, 200, 148),
                Color.rgb(237, 178, 100),
                Color.rgb(209, 142, 54),};
        float[] scales = new float[] {0.6F, 0.8F, 1};

        mWinds = new Wind[WIND_COUNT];
        for (int i = 0; i < mWinds.length; i ++) {
            mWinds[i] = new Wind(
                    canvasSizes[0],
                    canvasSizes[1],
                    colors[i * 3 / mWinds.length],
                    scales[i * 3 / mWinds.length]);
        }

        mLastDisplayRate = 0;
        mLastRotation3D = INITIAL_ROTATION_3D;
    }

    @Override
    public void updateData(@Size(2) int[] canvasSizes, long interval,
                           float rotation2D, float rotation3D) {
        for (Wind w : mWinds) {
            w.move(interval, mLastRotation3D == INITIAL_ROTATION_3D ? 0 : rotation3D - mLastRotation3D);
        }
        mLastRotation3D = rotation3D;
    }

    @Override
    public void draw(@Size(2) int[] canvasSizes, Canvas canvas,
                     float displayRate, float scrollRate, float rotation2D, float rotation3D) {

        if (displayRate >= 1) {
            canvas.drawColor(mBackgroundColor);
        } else {
            canvas.drawColor(
                    ColorUtils.setAlphaComponent(
                            mBackgroundColor,
                            (int) (displayRate * 255)));
        }

        if (scrollRate < 1) {
            rotation2D -= 16;
            canvas.rotate(
                    rotation2D,
                    canvasSizes[0] * 0.5F,
                    canvasSizes[1] * 0.5F);

            for (Wind w : mWinds) {
                mPaint.setColor(w.color);
                if (displayRate < mLastDisplayRate) {
                    mPaint.setAlpha((int) (displayRate * (1 - scrollRate) * 255));
                } else {
                    mPaint.setAlpha((int) ((1 - scrollRate) * 255));
                }
                canvas.drawRect(w.rectF, mPaint);
            }
        }

        mLastDisplayRate = displayRate;
    }

    @ColorInt
    public static int getThemeColor() {
        return Color.rgb(233, 158, 60);
    }
}

File: ui-weather-view/src/main/java/org/breezyweather/ui/theme/weatherView/materialWeatherView/implementor/WindImplementor.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.ui.theme.weatherView.materialWeatherView.implementor

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import androidx.annotation.ColorInt
import androidx.annotation.Size
import androidx.core.graphics.toColorInt
import org.breezyweather.ui.theme.weatherView.materialWeatherView.MaterialWeatherView.WeatherAnimationImplementor
import java.util.Random
import kotlin.math.cos
import kotlin.math.pow
import kotlin.math.sin

class WindImplementor(
    @Size(2) canvasSizes: IntArray,
    animate: Boolean,
    daylight: Boolean,
) : WeatherAnimationImplementor() {
    private val mAnimate = animate
    private val mPaint = Paint().apply {
        style = Paint.Style.FILL
        isAntiAlias = true
        alpha = ((if (daylight) 1f else 0.33f) * 255).toInt()
    }
    private val mWinds: Array<Wind>
    private var mLastRotation3D: Float

    private class Wind(
        private val mViewWidth: Int,
        private val mViewHeight: Int,
        @ColorInt val color: Int,
        val scale: Float,
    ) {
        var x = 0f
        var y = 0f
        var width = 0f
        var height = 0f
        var rectF: RectF = RectF()
        var speed: Float

        private val mCanvasSize: Int
        private val maxWidth: Float
        private val minWidth: Float
        private val maxHeight: Float
        private val minHeight: Float

        init {
            mCanvasSize = (mViewWidth * mViewWidth + mViewHeight * mViewHeight).toDouble().pow(0.5).toInt()
            speed = (mCanvasSize / (1000.0 * (0.5 + Random().nextDouble())) * 6.0).toFloat()
            maxHeight = 0.007f * mCanvasSize
            minHeight = 0.005f * mCanvasSize
            maxWidth = maxHeight * 10
            minWidth = minHeight * 6
            init(true)
        }

        private fun init(firstTime: Boolean) {
            val r = Random()
            y = r.nextInt(mCanvasSize).toFloat()
            x = if (firstTime) {
                (r.nextInt((mCanvasSize - maxHeight).toInt()) - mCanvasSize).toFloat()
            } else {
                -maxHeight
            }
            width = minWidth + r.nextFloat() * (maxWidth - minWidth)
            height = minHeight + r.nextFloat() * (maxHeight - minHeight)
            buildRectF()
        }

        private fun buildRectF() {
            val x = (x - (mCanvasSize - mViewWidth) * 0.5).toFloat()
            val y = (y - (mCanvasSize - mViewHeight) * 0.5).toFloat()
            rectF.set(x, y, x + width * scale, y + height * scale)
        }

        fun move(interval: Long, deltaRotation3D: Float) {
            x += speed
                .times(interval)
                .times(
                    scale.toDouble().pow(1.5) + 5 * sin(deltaRotation3D * Math.PI / 180.0) * cos(16 * Math.PI / 180.0)
                ).toFloat()
            y -= (speed * interval * 5 * sin(deltaRotation3D * Math.PI / 180.0) * sin(16 * Math.PI / 180.0))
                .toFloat()
            if (x >= mCanvasSize) {
                init(false)
            } else {
                buildRectF()
            }
        }
    }

    init {
        val colors = if (daylight) {
            intArrayOf(
                "#C2E4CA".toColorInt(), // Color.rgb(240, 200, 148),
                "#B2E0BA".toColorInt(), // Color.rgb(237, 178, 100),
                "#D2F0DA".toColorInt() // Color.rgb(209, 142, 54)
            )
        } else {
            intArrayOf(
                "#313E3A".toColorInt(), // Color.rgb(240, 200, 148),
                "#529B73".toColorInt(), // Color.rgb(237, 178, 100),
                "#638170".toColorInt() // Color.rgb(209, 142, 54)
            )
        }
        val scales = floatArrayOf(0.6f, 0.8f, 1f)
        mWinds = Array(WIND_COUNT) { i ->
            Wind(
                canvasSizes[0],
                canvasSizes[1],
                colors[i * 3 / WIND_COUNT],
                scales[i * 3 / WIND_COUNT]
            )
        }
        mLastRotation3D = INITIAL_ROTATION_3D
    }

    override fun updateData(
        @Size(2) canvasSizes: IntArray,
        interval: Long,
        rotation2D: Float,
        rotation3D: Float,
    ) {
        for (w in mWinds) {
            w.move(interval, if (mLastRotation3D == INITIAL_ROTATION_3D) 0f else rotation3D - mLastRotation3D)
        }
        mLastRotation3D = rotation3D
    }

    override fun draw(
        @Size(2) canvasSizes: IntArray,
        canvas: Canvas,
        scrollRate: Float,
        rotation2D: Float,
        rotation3D: Float,
    ) {
        var rotation2Dc = rotation2D
        if (scrollRate < 1) {
            rotation2Dc -= 16f
            canvas.rotate(
                rotation2D,
                canvasSizes[0] * 0.5f,
                canvasSizes[1] * 0.5f
            )
            for (w in mWinds) {
                mPaint.color = w.color
                mPaint.alpha = ((1 - scrollRate) * 255).toInt()
                canvas.drawRect(w.rectF, mPaint)
            }
        }
    }

    companion object {
        private const val INITIAL_ROTATION_3D = 1000f
        private const val WIND_COUNT = 160

        @ColorInt
        fun getThemeColor(daylight: Boolean): Int {
            return if (daylight) -0x15325d else -0x6a798b
        }
    }
}
